### PR TITLE
Adaptive lower trust level

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
This PR introduce **breaking** change:
- introduce `convergence` key in the `explore` session of the input script. This key provides the method of checking exploration convergence. Now the options are `fixed-levels` and `adaptive-lower`.

Main changes of this PR:
- implement the adaptive lower trust level 
- convergence is reached when the lower trust level does not change much over iterations. 
- remove TrustLevel
- new interface of `ExplorationReportTrustLevels`

### CODE REVIEW

File: dpgen2/exploration/report/report_adaptive_lower.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+    def failed_ratio(
+            self,
+            tag = None,
+    ):
+        return float(len(self.failed)) / float(self.nframes)
+
+    def accurate_ratio(
+            self,
+            tag = None,
```
Comment (njzjz) [inline dpgen2/exploration/report/report_adaptive_lower.py:256]: I don't quite understand what `tag` is used for.

File: dpgen2/exploration/report/report_adaptive_lower.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        max_nframes    int
+                The maximal number of frames of candidates.
+
+        Returns
+        -------
+        cand_frames   List[Tuple[int,int]]
+                Candidate frames. A list of tuples: [(traj_idx, frame_idx), ...]
+        """
+        self.candi_picked = [(ii[0], ii[1]) for ii in self.candi]
+        if max_nframes and max_nframes < len(self.candi_picked):
```
Comment (njzjz) [inline dpgen2/exploration/report/report_adaptive_lower.py:296]: I am not sure if `max_nframes = 0` is accepted. But to handle this situation, `if max_nframes and ...` should be `if max_nframes is not None and ...`.

File: dpgen2/exploration/report/report_adaptive_lower.py (hunk 1 above)
Comment (wanghan-iapcm) [reply dpgen2/exploration/report/report_adaptive_lower.py:256]: It is reserved for further use, e.g. we want to calculate the accurate ratio of a certain subgroup of the explored configurations.

File: dpgen2/exploration/report/report_adaptive_lower.py (hunk 2 above)
Comment (wanghan-iapcm) [reply dpgen2/exploration/report/report_adaptive_lower.py]: Right, fixed, thanks!

**Review by njzjz — APPROVED** [review]

**Pull request merged**